### PR TITLE
Homework #3: added code given to us, fixed failing tests and added a gui component

### PR DESCRIPTION
**Pull request by YaffaLevy**
Screen Shot 2024-03-19 at 4 33 44 PM

### CODE REVIEW

File: src/main/java/levy/fallingsand/Main.java [hunk 1]
```diff
@@ -5,6 +5,7 @@
     public class Main {
 
         public static void main(String[] args) {
+            new SandFrame().setVisible(true);
```
Comment (andrewoid) [inline src/main/java/levy/fallingsand/Main.java]: Do not put this into your command line program

**Review by andrewoid — COMMENTED** [review]
Post a screenshot

File: src/main/java/levy/fallingsand/Main.java (hunk 1 above)
Comment (YaffaLevy) [reply src/main/java/levy/fallingsand/Main.java]: what should I do instead

Comment (andrewoid) [reply src/main/java/levy/fallingsand/Main.java]: What you were instructed to do...

File: src/main/java/levy/fallingsand/Main.java [hunk 2]
```diff
@@ -5,6 +5,9 @@
     public class Main {
 
         public static void main(String[] args) {
+            //new SandFrame().setVisible(true);
+            SandFrame frame = new SandFrame();
+            frame.setVisible(true);
```
Comment (andrewoid) [inline src/main/java/levy/fallingsand/Main.java]: Like I said. 
Do not put this in your command line program.

File: src/main/java/levy/fallingsand/SandComponent.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    public SandComponent(Sand sand) {
+        this.sand = sand;
+    }
+
+    @Override
+    protected void paintComponent(Graphics g) {
+        super.paintComponent(g);
+
+        int cellWidth = getWidth() / sand.getWidth();
+        int cellHeight = getHeight() / sand.getHeight();
```
Comment (andrewoid) [inline src/main/java/levy/fallingsand/SandComponent.java:19]: This is good.

File: src/main/java/levy/fallingsand/Main.java (hunk 2 above)
Comment (YaffaLevy) [reply src/main/java/levy/fallingsand/Main.java]: it simply said create a gui so i did like we did for projectile. I am confused

Comment (andrewoid) [reply src/main/java/levy/fallingsand/Main.java]: I didn't say combine it with your command line program. Create a new class with a main

File: src/main/java/levy/fallingsand/Main.java [hunk 4]
```diff
@@ -2,20 +2,30 @@
...
-                System.out.println(sand);
-                System.out.println("Press Enter to make sand fall");
-                scanner.nextLine();
-                sand.fall();
-            }
+public class Main {
+
+    public static void main(String[] args) {
+
+        initializeGUI();
```
Comment (andrewoid) [inline src/main/java/levy/fallingsand/Main.java]: You are still not listening to what I'm telling you.

File: src/main/java/levy/fallingsand/Main.java [hunk 5]
```diff
@@ -2,20 +2,27 @@
...
+        SandComponent sandComponent = new SandComponent(sand);
+        sandComponent.initializeGUI();
+
+        while (true) {
+            System.out.println(sand);
+            System.out.println("Press Enter to make sand fall");
+            scanner.nextLine();
+            sand.fall();
+            sandComponent.repaint();
         }
```
Comment (andrewoid) [inline src/main/java/levy/fallingsand/Main.java]: You aren't understanding me.
Just get rid of the command line program.

File: src/main/java/levy/fallingsand/SandComponent.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+                    g.fillRect(x * cellWidth, y * cellHeight, cellWidth, cellHeight);
+                }
+            }
+        }
+    }
+
+    public void initializeGUI() {
+        SandFrame frame = new SandFrame();
+        frame.add(this);
+        frame.setVisible(true);
```
Comment (andrewoid) [inline src/main/java/levy/fallingsand/SandComponent.java]: This is not how we worked with frames before

File: src/main/java/levy/fallingsand/SandComponent.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+                    g.fillRect(x * cellWidth, y * cellHeight, cellWidth, cellHeight);
+                }
+            }
+        }
+    }
+
+    public void initializeGui() {
+        SandFrame frame = new SandFrame();
+        frame.setVisible(true);
+    }
```
Comment (andrewoid) [inline src/main/java/levy/fallingsand/SandComponent.java]: This is not how I showed you how to do this.

File: src/main/java/levy/fallingsand/Sand.java [hunk 8]
```diff
@@ -52,28 +58,41 @@ public void fall() {
...
-                    } else if (field[y + 1][x + direction2] == 0) {
-                        field[y][x] = 0;
-                        field[y + 1][x + direction2] = 1;
+                    // left edge
+                    if (x == 0) {
+                        if (field[y + 1][x] == 0) {
+                            field[y][x] = 0;
+                            field[y + 1][x] = 1;
+                        }
+                    }
```
Comment (andrewoid) [inline src/main/java/levy/fallingsand/Sand.java]: What does this code do from line 62-67?

Comment (YaffaLevy) [reply src/main/java/levy/fallingsand/Sand.java]: checks the left edge to see if anything there if nothing below it falls if something is below it stays

Comment (andrewoid) [reply src/main/java/levy/fallingsand/Sand.java]: If you're on the left edge, you should be able to fall to the right. Not just down.

File: src/main/java/levy/fallingsand/Sand.java [hunk 9]
```diff
@@ -52,28 +58,41 @@ public void fall() {
...
-                        field[y][x] = 0;
-                        field[y + 1][x + direction1] = 1;
-                    } else if (field[y + 1][x + direction2] == 0) {
-                        field[y][x] = 0;
-                        field[y + 1][x + direction2] = 1;
+                    // left edge
+                    if (x == 0) {
+                        if (field[y + 1][x] == 0) {
+                            field[y][x] = 0;
+                            field[y + 1][x + 1] = 1;
```
Comment (andrewoid) [inline src/main/java/levy/fallingsand/Sand.java:65]: You can't move a piece of sand to a space unless you know it's empty. That's what the if statement checks

**Pull request merged**